### ISSUE TITLE
Another case of 'Access Denied'

### ISSUE THREAD

**Issue by scratchpads** [NaturalHistoryMuseum/scratchpads2#6400]
- Added by [tsrsilva](https://strumigenys.myspecies.info/user/2)
- Scratchpad: [NeoStrumi](https://strumigenys.myspecies.info)

### Description:

I know that a similar issue occurred in the past (issue #5843), but I believe this is another recent case of 'Acces Denied'. I am having this issue when trying to select terms in a non-species vocabulary. Whenever I add a new vocabulary that is not a biological classification, all entries become inaccessible through the menu tab for me and other users. When we click in the name of the term in the vocabulary (for example, when I click on the term 'Continuant entity' in the Anatomical terms vocabulary in Neostrumi - https://strumigenys.myspecies.info/nontaxonomy/term/2571) the 'Access Denied' notice appears. I don't know if it's related to user's roles and/or permissions or something entirely different.

**Comment by therobyouknow** [comment]
I think you're right, this seems likely to be solved by that similar issue #5843 . I'll:
- set up a copy of the site in my development setup to reproduce the issue
- test the similar fix as in #5843 on this development setup. 
- If successful, apply this fix to your live setup to resolve the issue

**Comment by Archilegt** [comment]
This is a general problem. Access is also denied to children terms in the nested glossary of Myriatrix (https://myriatrix.myspecies.info/classification/14). An example of "Access denied" term is "scape" (https://myriatrix.myspecies.info/morphology/scape), nested in "Antenna".

**Comment by benscott** [comment]
Thanks @Archilegt & yep to confirm this is a recurring problem (see also #6380), where access to taxonomy terms is sometimes denied.

**Comment by therobyouknow** [comment]
@benscott are there particular steps to resolve the issue previously?

I'm aware that this issue then seems to come back, but would still want to be aware of any steps we can recall to see if they can be enhanced to avoid the issue re-occurring. Thank you!

**Comment by benscott** [comment]
I think if you clear the menu cache, that might resolve it?  And then somehow the access denied gets cached, and then becomes inaccessible?  

I *think* the culprit might be this module - https://github.com/NaturalHistoryMuseum/scratchpads2/tree/9387b91574857caf9b2b06bc89c5065068c41390/sites/all/modules/custom/scratchpads/scratchpads_species. It adds tabs to the species taxonomy pages - so changing the menu links for species vocabularies - but could be interferring with the non-species taxonomy tabs.  I'm not sure though, but could be a good place to start?

**Comment by therobyouknow** [comment]
Thank you Ben - yes - I can see opportunities in that code to be investigated with regard to permissions problems, and now that I can reproduce it locally on personal dev setup (see below), I can trace the logic.

Further to the above summary, the further info below in this comment is:
- just to confirm I can reproduce the original problem of "Access Denied" on my local dev setup. 
- to show how I reproduced the issue
- to outline next steps for action

Next steps:
- Which means I can now have xdebug help me step through the codebase to trace the logic.

Steps:
1 - Start to create a new taxonomy 'test' ( go to: admin/structure/taxonomy/add )
2 - On the page for admin/structure/taxonomy/add , select "Not a biological classification"

![image](https://user-images.githubusercontent.com/1080646/136062152-52254222-539d-429c-92b9-d9d09fbeb8bd.png)

3 - Click Save
4 - Then add some terms - admin/structure/taxonomy/test - click "Add term"
5 - These terms appear on the left after being added
6 - Hovering on a term shows the link to 'View' the term

![image](https://user-images.githubusercontent.com/1080646/136063033-e1747e8d-e2b2-4192-980a-bd96da5fcd69.png)


7 - Click on 'View' to view the term, then observe "Access Denied" - confirming able to reproduce original issue.


![image](https://user-images.githubusercontent.com/1080646/136061752-fc31478b-bdbc-42ca-8667-34bed34fccda.png)

**Comment by therobyouknow** [comment]
I'll setup something in the calendar for Monday @Archilegt - I'll email you an invite based on your online profiles.

**Comment by Archilegt** [comment]
Tomorrow Tuesday I am booked at 14:00-16:00 Helsinki time. Shall we try in the morning, at your 09:00, my 12:00?

**Comment by therobyouknow** [comment]
Solution available - PR for @benscott to review please: https://github.com/NaturalHistoryMuseum/scratchpads2/pull/6420/files

For example, this page - https://scratchpads-dev.ddev.site/morphology/forcipule

will now load and display (without the access denied)

AND

it just has the 'View' tab.

let me know if you need any other tabs to show - I can just add them to the code.

![image](https://user-images.githubusercontent.com/1080646/137314148-f458c021-7daa-4fdf-80a8-eac169e179f4.png)

**Comment by therobyouknow** [comment]
Released fix in 2.10.1 and deployed to [myriatrix.myspecies.info](https://myriatrix.myspecies.info/) 


Post release testing successful:

Including as per comment: https://github.com/NaturalHistoryMuseum/scratchpads2/issues/6400#issuecomment-931054896 these now work:

https://myriatrix.myspecies.info/nontaxonomy/term/37297

![image](https://user-images.githubusercontent.com/1080646/144397880-40813d0a-68b0-45d4-9058-f893ed90b2a9.png)


https://myriatrix.myspecies.info/morphology/scape

![image](https://user-images.githubusercontent.com/1080646/144399391-a6229cac-b352-4c9b-8085-6a3d2a00b493.png)


We'll roll out 2.10.1 to more sites today.

**Comment by therobyouknow** [comment]
The next site for the release will be:  https://strumigenys.myspecies.info/ so that the same issue will be fixed there, as described in first comment https://github.com/NaturalHistoryMuseum/scratchpads2/issues/6400#issue-977632010 

Following will be fixed once this release is complete today.  Will update when released.

https://strumigenys.myspecies.info/nontaxonomy/term/2571 - click on Anatomical entity, get access denied at page: https://strumigenys.myspecies.info/anatomical-terms/anatomical-entity


![image](https://user-images.githubusercontent.com/1080646/144401208-a1e53def-90eb-4782-98fb-8449d6ced1dd.png)

**Comment by therobyouknow** [comment]
Released 2.10.1 to trumigenys.myspecies.info - https://strumigenys.myspecies.info/nontaxonomy/term/2574 now works

![image](https://user-images.githubusercontent.com/1080646/144403522-c49e8726-00af-4962-89cd-fc878f718c94.png)

**Comment by Archilegt** [comment]
@therobyouknow, many thanks! I took a quick look and in Myriatrix it seems to be working as intended. 
Please, note that there are two types of lexicon display in Scratchpads. Myriatrix displays the lexicon with a classification tree on the left and the terms on the right. However, there is another display type programmed in Scratchpads, which shows one term and then a clickable list of alphabetically organized letters below, as a way of term navigation. That setting will need a separate check.
Also, please check with @benscott what was the code and settings for the autotagging that the terms in the glossaries were meant to enable. That seemed not to be working. Example: If I have the term "antenna" in the glossary and enable an autotagging setting, the word "antenna" should be recognized anywhere else in Myriatrix text, and autotagged with a link to the glossary term. It is a very important feature and it would be very nice to have it back. :)

**Comment by therobyouknow** [comment]
thank you @Archilegt for advising on Myriatrix working as intended.

With regard o your other items - I've read through and will come back to them after I've rolled out 2.10.1 release to other sites.

Can you expand on your points with some screenshots and URLs, advising on what to click to show the things on the site you need adjusting.  My guess is that we'd want to create a new github issue for these feature(s).

**Comment by therobyouknow** [comment]
will update https://phthiraptera.myspecies.info/ once I've located it to determine what commands to run. Usual commands give me 'not found'

**Comment by therobyouknow** [comment]
Closing this issue on the basis of your comment above https://github.com/NaturalHistoryMuseum/scratchpads2/issues/6400#issuecomment-984582351  I have raised 2 new issues based on your additional comments in that comment:

> @therobyouknow, many thanks! I took a quick look and in Myriatrix it seems to be working as intended.

> Please, note that there are two types of lexicon display in Scratchpads. Myriatrix displays the lexicon with a classification tree on the left and the terms on the right. However, there is another display type programmed in Scratchpads, which shows one term and then a clickable list of alphabetically organized letters below, as a way of term navigation. That setting will need a separate check.

- new issue raised for the above: https://github.com/NaturalHistoryMuseum/scratchpads2/issues/6573 - please advise if this still is an issue.

> Also, please check with @benscott what was the code and settings for the autotagging that the terms in the glossaries were meant to enable. That seemed not to be working. Example: If I have the term "antenna" in the glossary and enable an autotagging setting, the word "antenna" should be recognized anywhere else in Myriatrix text, and autotagged with a link to the glossary term. It is a very important feature and it would be very nice to have it back. :)

- new issue raised for the above: https://github.com/NaturalHistoryMuseum/scratchpads2/issues/6572 - please advise if this still is an issue.
